File: crystal/tests/utilities.py
###############################
#  This file is part of PyLaDa.
#
#  Copyright (C) 2013 National Renewable Energy Lab
# 
#  PyLaDa is a high throughput computational platform for Physics. It aims to make it easier to submit
#  large numbers of jobs on supercomputers. It provides a python interface to physical input, such as
#  crystal structures, as well as to a number of DFT (VASP, CRYSTAL) and atomic potential programs. It
#  is able to organise and launch computational jobs on PBS and SLURM.
# 
#  PyLaDa is free software: you can redistribute it and/or modify it under the terms of the GNU General
#  Public License as published by the Free Software Foundation, either version 3 of the License, or (at
#  your option) any later version.
# 
#  PyLaDa is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
#  Public License for more details.
# 
#  You should have received a copy of the GNU General Public License along with PyLaDa.  If not, see
#  <http://www.gnu.org/licenses/>.
###############################

""" Test some utilities from crystal extension module. """
def test_periodic():
  """ Test periodic images. """
  from pylada.crystal.cppwrappers import are_periodic_images
  from numpy import array, dot
  from numpy.linalg import inv
  from random import randint, random
  cell = array([ [0, 0.5, 0.5], [0.5, 0, 0.5], [0.5, 0.5, 0] ])
  invcell = inv(cell)
  for i in range(10):
    vec = dot(cell, array([random(), random(), random()]))
    for j in range(10):
      vec2 = vec + dot(cell, array([randint(-10, 11), randint(-10, 11), randint(-10, 11)]))
      assert are_periodic_images(vec, vec2, invcell)
      vec3 = vec2 + dot(cell, array([random()+0.0001, random(), random()]))
      assert not are_periodic_images(vec, vec3, invcell)

def test_into_cell():
  """ Test that vector is folded to fractional coordinates >= 0 and < 1. """
  from pylada.crystal.cppwrappers import are_periodic_images, into_cell
  from numpy import array, dot, all
  from numpy.linalg import inv
  from random import uniform
  cell = array([ [0, 0.5, 0.5], [0.5, 0, 0.5], [0.5, 0.5, 0] ])
  invcell = inv(cell)
  for i in range(100):
    vec = dot(cell, array([uniform(-10, 10), uniform(-10, 10), uniform(-10, 10)]))
    vec2 = into_cell(vec, cell, invcell)
    assert are_periodic_images(vec, vec2, invcell)
    assert all(dot(invcell, vec2) >= 0e0) and all(dot(invcell, vec2) < 1e0)
  
def test_zero_centered():
  """ Test that vector is folded to fractional coordinates >= -0.5 and < 0.5. """
  from pylada.crystal.cppwrappers import are_periodic_images, zero_centered
  from numpy import array, dot, all
  from numpy.linalg import inv
  from random import uniform
  cell = array([ [0, 0.5, 0.5], [0.5, 0, 0.5], [0.5, 0.5, 0] ])
  invcell = inv(cell)
  for i in range(100):
    vec = dot(cell, array([uniform(-10, 10), uniform(-10, 10), uniform(-10, 10)]))
    vec2 = zero_centered(vec, cell, invcell)
    assert are_periodic_images(vec, vec2, invcell)
    assert all(dot(invcell, vec2) >= -0.5) and all(dot(invcell, vec2) < 0.5)

def test_into_voronoi():
  """ Test that vector is folded into first Brillouin zone. """
  from pylada.crystal.cppwrappers import are_periodic_images, into_voronoi
  from numpy import array, dot
  from numpy.linalg import inv
  from random import uniform
  cell = array([ [0, 0.5, 0.5], [0.5, 0, 0.5], [0.5, 0.5, 0] ])
  invcell = inv(cell)
  for i in range(100):
    vec = dot(cell, array([uniform(-10, 10), uniform(-10, 10), uniform(-10, 10)]))
    vec2 = into_voronoi(vec, cell, invcell)
    assert are_periodic_images(vec, vec2, invcell)
    n = dot(vec2, vec2)
    assert dot(vec, vec) >= n
    for j in range(-3, 4):
      for k in range(-3, 4):
        for l in range(-3, 4):
          o = vec2 + dot(cell, [j, k, l])
          assert dot(o, o) >= n

if __name__ == "__main__":
  from sys import argv, path 
  if len(argv) > 0: path.extend(argv[1:])

  test_periodic()
  test_into_cell()
  test_zero_centered()
  test_into_voronoi()

File: crystal/utilities.cc
/******************************
   This file is part of PyLaDa.

   Copyright (C) 2013 National Renewable Energy Lab
  
   PyLaDa is a high throughput computational platform for Physics. It aims to make it easier to submit
   large numbers of jobs on supercomputers. It provides a python interface to physical input, such as
   crystal structures, as well as to a number of DFT (VASP, CRYSTAL) and atomic potential programs. It
   is able to organise and launch computational jobs on PBS and SLURM.
  
   PyLaDa is free software: you can redistribute it and/or modify it under the terms of the GNU General
   Public License as published by the Free Software Foundation, either version 3 of the License, or (at
   your option) any later version.
  
   PyLaDa is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
   the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
   Public License for more details.
  
   You should have received a copy of the GNU General Public License along with PyLaDa.  If not, see
   <http://www.gnu.org/licenses/>.
******************************/

math::rVector3d into_cell( math::rVector3d const &_vec, 
                           math::rMatrix3d const &_cell, 
                           math::rMatrix3d const &_inv)
{
  math::rVector3d result( _inv * _vec );
  result(0) -= std::floor(result(0)+types::roundoff);
  result(1) -= std::floor(result(1)+types::roundoff);
  result(2) -= std::floor(result(2)+types::roundoff);
  return _cell * result;
}

math::rVector3d zero_centered( math::rVector3d const &_vec, 
                               math::rMatrix3d const &_cell, 
                               math::rMatrix3d const &_inv)
{
  math::rVector3d result( _inv * _vec );
  result(0) -= std::floor(5e-1+result(0)+types::roundoff);
  result(1) -= std::floor(5e-1+result(1)+types::roundoff);
  result(2) -= std::floor(5e-1+result(2)+types::roundoff);
  // numerical stability check.
  if( math::eq(result(0), 5e-1) ) result(0) = -5e-1;
  else if( math::lt(result(0), -5e-1)) result(0) += 1e0;
  if( math::eq(result(1), 5e-1) ) result(1) = -5e-1;
  else if( math::lt(result(1), -5e-1)) result(1) += 1e0;
  if( math::eq(result(2), 5e-1) ) result(2) = -5e-1;
  else if( math::lt(result(2), -5e-1)) result(2) += 1e0;
  return _cell * result;
}

math::rVector3d into_voronoi( math::rVector3d const &_vec, 
                              math::rMatrix3d const &_cell, 
                              math::rMatrix3d const &_inv)
{
  math::rVector3d result( _inv * _vec );
  result(0) -= std::floor(result(0)+types::roundoff);
  result(1) -= std::floor(result(1)+types::roundoff);
  result(2) -= std::floor(result(2)+types::roundoff);
  // numerical stability check.
  if( math::eq(result(0), -1e0) or math::eq(result(0), 1e0) ) result(0) = 0e0;
  if( math::eq(result(1), -1e0) or math::eq(result(1), 1e0) ) result(1) = 0e0;
  if( math::eq(result(2), -1e0) or math::eq(result(2), 1e0) ) result(2) = 0e0;
  math::rVector3d const orig(result);
  types::t_real min_norm = (_cell*orig).squaredNorm();
  for(int i(-1); i < 2; ++i)
    for(int j(-1); j < 2; ++j)
      for(int k(-1); k < 2; ++k)
      {
        math::rVector3d const translated = orig + math::rVector3d(i,j,k);
        types::t_real const d( (_cell*translated).squaredNorm() );
        if( math::gt(min_norm, d) )
        {
          min_norm = d;
          result = translated;
        }
      }
  return _cell * result;
}
